### PR TITLE
Person Info Panel Fix

### PR DESCRIPTION
**Pull request by nbriannl**
Edit-tag/Delete-tag, Delete Person and Clear does not update person info panel when viewing a person that has it's tag modified or is the one being deleted.

This PR fixes this.

Refer to my code review below for an explanation on the fix.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -19,3 +19,4 @@ classes/
 src/main/resources/docs/
 src/main/resources/images/contactPhotos
 out/
+Collate-TUI.jar
```
Comment (nbriannl) [inline .gitignore:22]: Added to gitignore because I accidentally pushed Collate-TUI into the repo.

File: src/main/java/seedu/address/commons/events/ui/PersonHasBeenDeletedEvent.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+import seedu.address.commons.events.BaseEvent;
+import seedu.address.model.person.ReadOnlyPerson;
+
+//@@author nbriannl
+
+/**
+ * Represents a deletion of a person.
+ */
+public class PersonHasBeenDeletedEvent extends BaseEvent {
```
Comment (nbriannl) [inline src/main/java/seedu/address/commons/events/ui/PersonHasBeenDeletedEvent.java:11]: `PersonHasBeenDeletedEvent` `PersonHasBeenModifiedEvent` are events Person Info Panel will look out for.

File: src/main/java/seedu/address/model/AddressBook.java [hunk 3]
```diff
@@ -211,7 +215,12 @@ public boolean removePerson(ReadOnlyPerson key) throws PersonNotFoundException {
             removeContactPhoto(photoPath);
         }
 
-        return persons.remove(key);
+        if (persons.remove(key)) {
+            EventsCenter.getInstance().post(new PersonHasBeenDeletedEvent(key));
```
Comment (nbriannl) [inline src/main/java/seedu/address/model/AddressBook.java:219]: Only raise the event on a successful deletion

Comment (nbriannl) [reply src/main/java/seedu/address/model/AddressBook.java:219]: Raising event is on an AddressBook level as oppposed to a Model level.

It is much more clearer that you raise an event at the exact moment a **successful** deletion occurs as opposed to raising an event in all methods that involve a deletion of a person.
Similar to updating a person (during edit tag or delete tag)

File: src/main/java/seedu/address/ui/PersonInfoPanel.java [hunk 4]
```diff
@@ -31,6 +34,7 @@
     private static String DEFAULT_PHOTO_PATH = "/images/defaultPhoto.jpg";
 
     private ReadOnlyPerson person;
+    private ReadOnlyPerson currentlyViewedPerson;
```
Comment (nbriannl) [inline src/main/java/seedu/address/ui/PersonInfoPanel.java:37]: PersonInfoPanel keeps tracks of the currentViewedPerson in the the App

File: src/main/java/seedu/address/ui/PersonInfoPanel.java [hunk 5]
```diff
@@ -76,9 +80,11 @@ private void loadDefaultPerson() {
         email.setText("");
         birthday.setText("");
         remark.setText("");
-
+        tags.getChildren().clear();
 
         setDefaultContactPhoto();
+        currentlyViewedPerson = null;
+        logger.info("Currently Viewing: Default Person" );
     }
```
Comment (nbriannl) [inline src/main/java/seedu/address/ui/PersonInfoPanel.java:88]: it is null when the default person is loaded

File: src/main/java/seedu/address/ui/PersonInfoPanel.java [hunk 6]
```diff
@@ -103,8 +109,27 @@ private void loadPerson(ReadOnlyPerson person) {
         });
 
         loadPhoto(person);
+
+        currentlyViewedPerson = person;
+        logger.info("Currently Viewing: " + currentlyViewedPerson.getName() );
+    }
```
Comment (nbriannl) [inline src/main/java/seedu/address/ui/PersonInfoPanel.java:115]: And a person is assigned to be the currentlyViewedPerson when it is loaded in the app.

File: src/main/java/seedu/address/model/AddressBook.java [hunk 7]
```diff
@@ -120,6 +123,7 @@ public void updatePerson(ReadOnlyPerson target, ReadOnlyPerson editedReadOnlyPer
         // This can cause the tags master list to have additional tags that are not tagged to any person
         // in the person list.
         persons.setPerson(target, editedPerson);
+        EventsCenter.getInstance().post(new PersonHasBeenModifiedEvent(target, editedPerson));
```
Comment (nbriannl) [inline src/main/java/seedu/address/model/AddressBook.java:126]: Only raise the event on a successful modification (i.e. no exception thrown)

File: src/main/java/seedu/address/ui/PersonInfoPanel.java [hunk 8]
```diff
@@ -152,17 +177,43 @@ public void loadPhoto(ReadOnlyPerson person) {
...
+    //@@author nbriannl
+    @Subscribe
+    private void handleAddressBookChangedEvent(AddressBookChangedEvent event) {
+        logger.info(LogsCenter.getEventHandlingLogMessage(event));
+        if (event.data.getPersonList().size() == 0 && event.data.getTagList().size() == 0) {
+            clearBind();
+            loadDefaultPerson();
+        }
+    }
+
```
Comment (nbriannl) [inline src/main/java/seedu/address/ui/PersonInfoPanel.java:215]: The 3 events that trigger this, and the app will look out for:
- Person has been modified via Edit or Delete Tag
-- Check if the modified person is the currentlyViewedPerson
-- Then load the new person.
- Person has been deleted
-- Check if the deleted person is the currentlyViewedPerson
-- Then load a default blank person
- AddressBook has been changed
-- Check if the address book is empty
-- Then load a default person

**Review by nbriannl — COMMENTED** [review]
Since Travis is down. Here is a code review to explain the changes that fixes the Person Info Panel response to corner cases. 

Please test it on your end and raise any issues.
